src/cylinder.jl

```diff
@@ -25,12 +25,21 @@ function Base.in(x::SVector{N}, s::Cylinder{N}) where {N}
     return sum(abs2,d - p*s.a) ≤ s.r^2
 end
 
-function normal(x::SVector{N}, s::Cylinder{N}) where {N}
+function surfpt_nearby(x::SVector{N}, s::Cylinder{N}) where {N}
     d = x - s.c
-    p = dot(d, s.a)
-    p > s.h2 && return s.a
-    p < -s.h2 && return -s.a
-    return normalize(d - p*s.a)
+    p = d ⋅ s.a
+    q = d - p*s.a
+    lp = abs(p)
+    lq = norm(q)
+    if abs(lp-s.h2) < abs(lq-s.r)
+        nout = sign(p)*s.a
```

Comment:
(This could slightly break the symmetry by picking an arbitrary normal vector for points exactly in the center, but there are lots of other places in the code that do that too, e.g. if a point is exactly on the boundary between two objects, and it's too painful to track them all down.  In practice, libctl does the same thing and it doesn't cause a problem.)